modules/python/python-types.c

```diff
@@ -0,0 +1,439 @@
+  const gchar *charset;
+
+  if (len < 0)
+    len = strlen(value);
+
+  /* NOTE: g_get_charset() returns if the current character set is utf8 */
+  if (g_get_charset(&charset))
+    {
+      obj = PyUnicode_FromStringAndSize(value, len);
+    }
+  else
+    {
+      GError *error = NULL;
+      gsize bytes_read, bytes_written;
+      gchar *utf8_string;
+
+      utf8_string = g_locale_to_utf8(value, len, &bytes_read, &bytes_written, &error);
+      if (utf8_string)
+        {
+          obj = PyUnicode_FromStringAndSize(utf8_string, bytes_written);
+          g_free(utf8_string);
+        }
+      else
+        {
+          g_error_free(error);
+          obj = PyBytes_FromStringAndSize(value, len);
+        }
+    }
+
+  if (!obj)
+    _log_py_object_construction_error();
+
+  return obj;
+}
+
+PyObject *
+py_long_from_long(gint64 l)
+{
+  PyObject *obj = PyLong_FromLong(l);
+  if (!obj)
+    _log_py_object_construction_error();
+
+  return obj;
+}
+
+PyObject *
+py_double_from_double(gdouble d)
+{
+  PyObject *obj = PyFloat_FromDouble(d);
+  if (!obj)
+    _log_py_object_construction_error();
+
+  return obj;
+}
+
+PyObject *
+py_boolean_from_boolean(gboolean b)
+{
+  PyObject *obj = PyBool_FromLong(b);
+  if (!obj)
+    _log_py_object_construction_error();
+
+  return obj;
+}
+
+PyObject *
+py_list_from_list(const gchar *list, gssize list_len)
+{
+  PyObject *obj = PyList_New(0);
+  if (!obj)
+    {
+      _log_py_object_construction_error();
+      return NULL;
+    }
+
+  ListScanner scanner;
+  list_scanner_init(&scanner);
+  list_scanner_input_string(&scanner, list, list_len);
+
+  while (list_scanner_scan_next(&scanner))
+    {
+      PyObject *element = py_bytes_from_string(list_scanner_get_current_value(&scanner),
+                                               list_scanner_get_current_value_len(&scanner));
+      if (!element || PyList_Append(obj, element) != 0)
+        {
+          list_scanner_deinit(&scanner);
+          _log_py_object_construction_error();
+          Py_DECREF(obj);
+          return NULL;
+        }
+    }
+
+  list_scanner_deinit(&scanner);
+
+  return obj;
+}
+
+PyObject *
+py_obj_from_log_msg_value(const gchar *value, gssize value_len, LogMessageValueType type)
+{
+  switch (type)
+    {
+    case LM_VT_INTEGER:
+    {
+      gint64 l;
+      if (type_cast_to_int64(value, &l, NULL))
+        return py_long_from_long(l);
+      return NULL;
+    }
+
+    case LM_VT_DOUBLE:
+    {
+      gdouble d;
+      if (type_cast_to_double(value, &d, NULL))
+        return py_double_from_double(d);
+      return NULL;
+    }
+
+    case LM_VT_BOOLEAN:
+    {
+      gboolean b;
+      if (type_cast_to_boolean(value, &b, NULL))
+        return py_boolean_from_boolean(b);
+      return NULL;
+    }
+
+    case LM_VT_NULL:
+      Py_RETURN_NONE;
+
+    case LM_VT_LIST:
+      return py_list_from_list(value, value_len);
+
+    case LM_VT_STRING:
+    default:
+      return py_bytes_from_string(value, value_len);
```

Comment:
we can't convert JSON and DATETIME types at the moment. I am not sure we need or how we would do that, I just want to point it out.

This PR is already huge, so I wouldn't want to add those, we should just think it through if we need to add these now (e.g. 4.0) or we can do this later as well.

Reply:
Yes, I think we should not implement that in this PR. Let's talk about this on Thrusday.

Reply:
I think we could expose JSON values either by:

1) parsing the JSON using json.loads()
2) exposing it as a an str subclass instance (e.g. JSONValue or something like that that derives from str). This way we wouldn't have to parse the JSON value but would make it possible to detect that it was originally a JSON text and could allow Python to do the same.

In the case of DATETIME I think that's simpler as converting those to Datetime objects doesn't sound as heavyweight as parsing the JSON payload. I am filing this as a discussion item, let's fork it off from this PR.